org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/scopes/BuildScopeUnit.scala

```diff
@@ -0,0 +1,82 @@
+package org.scalaide.core.internal.project.scopes
+
+import scala.tools.nsc.Settings
+import org.eclipse.core.resources.IContainer
+import org.eclipse.core.resources.IFile
+import org.eclipse.core.resources.IResource
+import org.eclipse.core.runtime.IProgressMonitor
+import org.eclipse.core.runtime.SubMonitor
+import org.eclipse.jdt.core.IJavaModelMarker
+import org.scalaide.core.IScalaProject
+import org.scalaide.core.SdtConstants
+import org.scalaide.core.internal.builder.BuildProblemMarker
+import org.scalaide.core.internal.builder.zinc.SbtAwareBuildManager
+import org.scalaide.core.internal.project.CompileScope
+import sbt.inc.IncOptions
+import org.scalaide.core.internal.builder.zinc.EclipseSbtBuildManager
+
+class BuildScopeUnit(val scope: CompileScope, val owningProject: IScalaProject, settings: Settings,
+  val dependentUnitInstances: Seq[BuildScopeUnit] = Seq.empty)
+    extends SbtAwareBuildManager {
+
+  private val delegate =
+    new EclipseSbtBuildManager(owningProject, settings, Some(owningProject.underlying.getFile(".cache-" + scope.name)),
+        addToClasspath, srcOutputs)
+
+  private def isSrcOutPossessor: PartialFunction[(IContainer, IContainer), Boolean] = {
+    case (src, out) if scope.isValidSourcePath(src.getProjectRelativePath) => true
+    case _ => false
+  }
+
+  private def addToClasspath = owningProject.sourceOutputFolders.filterNot(isSrcOutPossessor) map {
+    case (src, out) => out.getLocation
+  }
+
+  private def srcOutputs = owningProject.sourceOutputFolders.filter(isSrcOutPossessor)
+
+  def sources: Seq[IContainer] = srcOutputs.map {
+    case (src, out) => src
+  }
+
+  def clean(implicit monitor: IProgressMonitor) = delegate.clean
+
+  override def build(addedOrUpdated: Set[IFile], removed: Set[IFile], monitor: SubMonitor): Unit = {
+    hasInternalErrors = if (areDependedUnitsBuilt) {
+      def getValidJavaSourcesOfThisScope: Set[IFile] = {
+        val Dot = 1
+        val javaSources = getToCompileFor(owningProject.allSourceFiles
+          .filter { _.getLocation.getFileExtension == SdtConstants.JavaFileExtn.drop(Dot) })
+        javaSources -- removed
+      }
+      def resetJavaMarkers(javaFiles: Set[IFile]): Set[IFile] = {
+        javaFiles.foreach { _.deleteMarkers(IJavaModelMarker.JAVA_MODEL_PROBLEM_MARKER, true, IResource.DEPTH_INFINITE) }
```

Comment:
problem here concerns situation when in scala file the method name/signature has been changed and this change should be visible to java files, so it justifies markers cleaning and adding files for compilation. java compiler should decide then if these files need or not to be rebuilt. anyway I double check it.